include/pybind11/pybind11.h

```diff
@@ -3148,17 +3148,22 @@ typing::Iterator<ValueType> make_value_iterator(Type &value, Extra &&...extra) {
 
 template <typename InputType, typename OutputType>
 void implicitly_convertible() {
+    static int tss_sentinel_pointee = 1; // arbitrary value
     struct set_flag {
-        bool &flag;
-        explicit set_flag(bool &flag_) : flag(flag_) { flag_ = true; }
-        ~set_flag() { flag = false; }
+        thread_specific_storage<int> &flag;
+        explicit set_flag(thread_specific_storage<int> &flag_) : flag(flag_) {
+            flag = &tss_sentinel_pointee; // trick: the pointer itself is the sentinel
+        }
+        ~set_flag() { flag.reset(nullptr); }
+
+        // Prevent copying/moving to ensure RAII guard is used safely
+        set_flag(const set_flag &) = delete;
+        set_flag(set_flag &&) = delete;
+        set_flag &operator=(const set_flag &) = delete;
+        set_flag &operator=(set_flag &&) = delete;
     };
     auto implicit_caster = [](PyObject *obj, PyTypeObject *type) -> PyObject * {
-#ifdef Py_GIL_DISABLED
-        thread_local bool currently_used = false;
-#else
-        static bool currently_used = false;
-#endif
+        auto &currently_used = detail::get_implicit_caster_recursion_guard();
```

Comment:
I'm not 100% sure, but I think this changes semantics. As I understand it:

* Previously, there was one `currently_used` variable per template instantiation. For example, `implicitly_convertible<int, double>` would have a different `bool currently_used` (static or thread_local) than `implicitly_convertible<float, int>`.
* Now, the guard is shared across the library via `get_implicit_caster_recursion_guard()`

Reply:
Yes, you're right.  I'm not sure whether the old behavior was part of the intent of the guard or not....  I guess there was not a test for this, or it would've failed at some point here....

Reply:
Thanks @colesbury for catching this!

> I guess there was not a test for this, or it would've failed at some point here....

For easy reference, this is when the runaway recursion guard + test was added:

https://github.com/pybind/pybind11/pull/1037/files

The test is only for one type, one thread.

I'm wondering if this actually correct in the general case (e.g. could `implicitly_convertible<int, A>` be in flight in two threads at the same time?) ... but that's something for another day, definitely not this PR.